File: git-epic-creator/services/code_graph_ingestion_service/src/core/stable_ids.py
"""Deterministic stable IDs and snippet hashing."""

from __future__ import annotations

import hashlib


def stable_node_id(*parts: str, size: int = 24) -> str:
    """Return a stable node_id derived from semantic identity + span."""

    raw = "|".join(parts).encode("utf-8", errors="replace")
    return hashlib.sha256(raw).hexdigest()[:size]


def _normalize_text_for_snippet_hash(text: str) -> str:
    """Normalize text for snippet hashing.

    Allowed normalizations (IMPLEMENTATION.md §4.2):
    - newline normalization (\r\n/\r -> \n)
    - trimming trailing whitespace
    """

    t = text.replace("\r\n", "\n").replace("\r", "\n")
    lines = t.split("\n")
    lines = [ln.rstrip(" \t") for ln in lines]
    return "\n".join(lines)


def snippet_hash(*, file_path: str, start_line: int, end_line: int, text: str) -> str:
    """Compute deterministic snippet hash for evidence fidelity."""

    norm = _normalize_text_for_snippet_hash(text)
    raw = f"{file_path}|{start_line}|{end_line}|\n{norm}".encode("utf-8", errors="replace")
    return hashlib.sha256(raw).hexdigest()

File: git-epic-creator/services/code_graph_ingestion_service/src/core/stable_ids.rs
use sha2::{Digest, Sha256};

pub fn stable_node_id<I, S>(parts: I) -> String
where
    I: IntoIterator<Item = S>,
    S: AsRef<str>,
{
    stable_node_id_with_size(parts, 24)
}

pub fn stable_node_id_with_size<I, S>(parts: I, size: usize) -> String
where
    I: IntoIterator<Item = S>,
    S: AsRef<str>,
{
    let raw = parts
        .into_iter()
        .map(|s| s.as_ref().to_string())
        .collect::<Vec<_>>()
        .join("|");

    let mut h = Sha256::new();
    h.update(raw.as_bytes());
    let digest = h.finalize();
    hex::encode(digest)[..size].to_string()
}

fn normalize_text_for_snippet_hash(text: &str) -> String {
    let t = text.replace("\r\n", "\n").replace('\r', "\n");
    let lines = t.split('\n').map(|ln| ln.trim_end_matches([' ', '\t']));
    lines.collect::<Vec<_>>().join("\n")
}

pub fn snippet_hash(file_path: &str, start_line: i64, end_line: i64, text: &str) -> String {
    let norm = normalize_text_for_snippet_hash(text);
    let raw = format!("{file_path}|{start_line}|{end_line}|\n{norm}");

    let mut h = Sha256::new();
    h.update(raw.as_bytes());
    hex::encode(h.finalize())
}


